include/at_cmd.h

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#ifndef AT_CMD_INTERFACE_H_
+#define AT_CMD_INTERFACE_H_
+
+/**
+ * @file at_cmd_interface.h
+ *
+ * @defgroup at_cmd_interface AT Command interface driver
+ *
+ * @{
+ *
+ * @brief Public APIs for the AT command interface driver.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+
+/**
+ * @brief AT command return codes
+ */
+enum at_cmd_rc {
+	AT_CMD_OK,
+	AT_CMD_ERROR,
+	AT_CMD_CMS,
+	AT_CMD_CME,
+	AT_CMD_NOTIFICATION,
+};
+
+/**
+ * @typedefs at_cmd_handler_t
+ *
+ * Because this driver let multiple threads share the same socket, it must make
+ * sure that the correct thread gets the correct data returned from the AT
+ * interface. The_cmd_write() function let the user specify the handler that
+ * will process the data from a specific AT command call. Notifications will be
+ * handled by the global handler defined using the at_cmd_set_handler()
+ * function. Both handlers are of the type at_cmd_handler_t.
+ *
+ * @param response     Null terminated string containing the modem message
+ *
+ */
+typedef void (*at_cmd_handler_t)(char *response);
+
+/**
+ * @brief Function to send an AT command to the modem
+ *
+ * @param cmd     Pointer to null terminated AT command string
+ * @param handler Pointer to handler that will process any returned data.
+ *                NULL pointer is allowed, which means that any returned data
+ *                will be discarded.
+ *
+ * @return One of the possible return codes defined by the @ref at_cmd_rc
+ *         enumerator
+ */
+enum at_cmd_rc at_cmd_write(const char *const cmd,
+			    at_cmd_handler_t handler);
+
+/**
+ * @brief Function to send an AT command and receive response immediately
+ *
+ * This function can be used for AT commands that return a response
+ * immediately. The response will returned in the supplied buffer.
+ * This function will return an empty buffer if nothing is returned by the
+ * AT command. If the response buffer is not large enough to hold the
+ * complete response it will be dropped, and an error will be logged.
+ *
+ * @param cmd Pointer to null terminated AT command string
+ * @param buf Buffer to put the response in
+ * @param buf_len Length of response buffer
+ *
+ * @return One of the possible return codes defined by the @ref at_cmd_rc
+ *         enumerator
+ */
+enum at_cmd_rc at_cmd_write_with_response(const char *const cmd,
+					  char *buf,
+					  size_t buf_len);
+
+/**
+ * @brief Function to set AT command global notification handler
+ *
+ * @param handler Pointer to a received notification handler function
+ */
+void at_cmd_set_handler(at_cmd_handler_t handler);
```

Comment:
`at_cmd_set_notification_handler` perhaps? It'd be easier to get the purpose of this function when reading the code.

Reply:
Yes, I kind of agree, the current name is confusing, but at_cmd_set_notification_handler is so long.

Reply:
Well, I prefer long names over cryptic ones, but perhaps it's me. No strong push on the change.

Reply:
No, I agree, I change it. The current one doesn't reveal it's true nature.